lnwallet/fee_estimator.go

```diff
@@ -458,3 +475,270 @@ func (b *BitcoindFeeEstimator) fetchEstimate(confTarget uint32) (SatPerKWeight,
+
+	// defaultFeePerKw is a fallback value that we'll use if we're unable
+	// to query the API for any reason.
+	defaultFeePerKw SatPerKWeight
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewWebAPIFeeEstimator creates a new WebAPIFeeEstimator from a given URL and a
+// fallback default fee. The fees are updated whenever a new block is mined.
+func NewWebAPIFeeEstimator(
+	api string, notifier chainntnfs.ChainNotifier, defaultFee SatPerKWeight) *WebAPIFeeEstimator {
+
+	return &WebAPIFeeEstimator{
+		apiURL:           api,
+		chainNotifier:    notifier,
+		feeByBlockTarget: [MaxBlockTarget + 1]uint32{},
+		defaultFeePerKw:  defaultFee,
+		quit:             make(chan struct{}),
+	}
+}
+
+// EstimateFeePerKW takes in a target for the number of blocks until an initial
+// confirmation and returns the estimated fee expressed in sat/kw.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) EstimateFeePerKW(numBlocks uint32) (SatPerKWeight, error) {
+	if numBlocks > MaxBlockTarget {
+		numBlocks = MaxBlockTarget
+	}
+
+	w.feesMtx.RLock()
+	feePerKb := w.feeByBlockTarget[numBlocks]
+	w.feesMtx.RUnlock()
+
+	// If the result is too low, then we'll clamp it to our current fee
+	// floor.
+	satPerKw := SatPerKVByte(feePerKb).FeePerKWeight()
+	if satPerKw < FeePerKwFloor {
+		satPerKw = FeePerKwFloor
+	}
+
+	walletLog.Debugf("Web API returning %v sat/kw for conf target of %v",
+		int64(satPerKw), numBlocks)
+
+	return satPerKw, nil
+}
+
+// Start signals the FeeEstimator to start any processes or goroutines it needs
+// to perform its duty.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Start() error {
+	if !atomic.CompareAndSwapUint32(&w.started, 0, 1) {
+		return nil
+	}
+
+	if err := w.chainNotifier.Start(); err != nil {
+		return err
+	}
+
+	blockEpochs, err := w.chainNotifier.RegisterBlockEpochNtfn(nil)
+	if err != nil {
+		return err
+	}
+	w.blockEpochs = blockEpochs
+
+	w.wg.Add(1)
+	go w.blockEpochHandler()
+
+	return nil
+}
+
+// Stop stops any spawned goroutines and cleans up the resources used by the
+// fee estimator.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Stop() error {
+	if !atomic.CompareAndSwapUint32(&w.stopped, 0, 1) {
+		return nil
+	}
+
+	close(w.quit)
+	w.wg.Wait()
+
+	return nil
+}
+
+// RelayFeePerKW returns the minimum fee rate required for transactions to be
+// relayed.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) RelayFeePerKW() SatPerKWeight {
+	return FeePerKwFloor
+}
+
+// parseAPIResponse attempts to parse the body of the response generated by the
+// API, into the JSON format: { fee_by_block_target: { a: b, c: d, ... y: z }}
+// where the keys are block targets and the values are satPerKB fee estimates.
+func (w *WebAPIFeeEstimator) parseAPIResponse(r io.Reader) (map[uint32]uint32, error) {
```

Comment:
Test for this would be nice 👍

Reply:
fixed :) the testing could still be stronger though. i'm still thinking about how (might add a ChainNotifier harness and an HTTP API harness...?)

Reply:
Can mock out the `ChainNotifier` interface (just send blocks), then also mock out the HTTP request itself since we can use the interface to just return the value directly.

Reply:
done :)